filebeat/input/file/comparator.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	cmpType := cmpTypes[0]
+	logp.Info(common.DebugString(c, false))
+	f, ok := comparatorFactories[cmpType]
+	if !ok {
+		return nil, fmt.Errorf("no such file_identity comparator: %s", cmpType)
+	}
+
+	c, err := c.Child(cmpType, -1)
+	if err != nil {
+		return nil, fmt.Errorf("error while transforming file_identity configuration")
+	}
+	return f(c)
+}
+
+// DefaultComparatorConfig returns a configuration with default values.
+func DefaultComparatorConfig() *common.Config {
+	return common.MustNewConfigFrom(map[string]interface{}{DefaultComparatorName: nil})
+}
+
+// inodeDeviceIDComparator compares files based on inode and device id.
+type inodeDeviceIDComparator struct {
+	log *logp.Logger
+}
+
+func newINodeDevideIDComparator(_ *common.Config) (IdentityComparator, error) {
+	return &inodeDeviceIDComparator{
+		log: logp.NewLogger(inodeDeviceIDName + comparatorNameSuffix),
+	}, nil
+}
+
+// IsSameFile checks if two states point to the same file
+// based on inode and device id.
+func (i *inodeDeviceIDComparator) IsSameFile(one *State, other *State) bool {
+	isSame := one.FileStateOS.IsSame(other.FileStateOS)
+	if isSame {
+		i.log.Debugf("two states correspond to the same file based on os.SameFile")
+		return true
+	}
+	i.log.Debugf("two states correspond different files based on os.SameFile")
+	return false
+}
+
+// IsEmptyState checks if a state corresponds to a file.
+func (i *inodeDeviceIDComparator) IsEmptyState(s *State) bool {
+	return s.FileStateOS == file.StateOS{} &&
+		s.Source == "" &&
+		len(s.Meta) == 0 &&
+		s.Timestamp.IsZero()
+}
+
+func (i *inodeDeviceIDComparator) StateIdentifier() StateIdentifier {
+	return &inodeDeviceIdentifier{name: inodeDeviceIDName}
+}
+
+func (i *inodeDeviceIDComparator) Name() string { return inodeDeviceIDName }
+
+// pathComparator compares files based on their paths.
+type pathComparator struct {
+	log *logp.Logger
+}
+
+func newPathComparator(_ *common.Config) (IdentityComparator, error) {
+	return &pathComparator{
+		log: logp.NewLogger(pathName + comparatorNameSuffix),
+	}, nil
+}
+
+// IsSameFile checks if the two paths correspond to the same file.
+func (p *pathComparator) IsSameFile(one *State, other *State) bool {
+	isSame := one.Source == other.Source
+	if isSame {
+		p.log.Debugf("two states correspond to the same file based on path names")
+		return true
+	}
+	p.log.Debugf("two states correspond different files based on path names")
+	return false
+}
+
+func (p *pathComparator) IsEmptyState(s *State) bool {
+	return s.Source == "" && len(s.Meta) == 0 && s.Timestamp.IsZero()
+}
+
+func (p *pathComparator) StateIdentifier() StateIdentifier {
+	return &pathIdentifier{name: pathName}
+}
+
+func (p *pathComparator) Name() string { return pathName }
+
+// inodeMarkerComparator compares files based on the inode and the contents of
+// a selected marker file.
+type inodeMarkerComparator struct {
+	log    *logp.Logger
+	config *inodeMarkerComparatorConfig
```

Comment:
the Config isn't really shared between types. No need to make it a pointer here.